text/0000-override-gem-dependency-versions.md

```diff
@@ -0,0 +1,123 @@
+end
+```
+
+This will allow bundler to install the `2.1.0` version of the `json` gem.
+
+The output when running `bundle install` will look like the following:
+
+```
+Installing json 2.1.0 (forced version, conflicts with other_gem)
+Installing other_gem 1.0.0
+```
+
+Most of the time, this should only be a temporary fix as we expect the gem author to release a new version with updated dependency constraints.
+
+To provide additional information a note can be included:
+
+```ruby
+gem 'json'
+gem 'other_gem' do |s|
+  s.override_runtime_dependency('json', '>= 1.7.7', '<3', 'waiting on pull request #5')
+end
+```
+
+The output when running `bundle install` will look like the following:
+
+```
+Installing json 2.1.0 (forced version, conflicts with other_gem: waiting on pull request #5)
+Installing other_gem 1.0.0
+```
+
+# Reference-level explanation (TODO)
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+# Drawbacks
+
+These are the common drawbacks and counter-arguments mentioned by the community:
+* This issue can already be solved by forking gem repositories and updating the dependencies yourself
+* Allowing this means there is no longer a canonical source of truth about what a gem's dependencies are
+* When these kinds of changes cause resolver errors (and they will), it will be extremely hard to diagnose
+* The potential that people will complain that Bundler is broken when it’s actually something they did themselves
+* The potential that people will complain to the gem author that the gem is broken when it’s actually something they did themselves
+
+These are all valid points of view, however, there are also counter-arguments to each of these points which are also equally valid:
+* The most obvious place to manage version dependencies (even overridden version constraints) is in the Gemfile and not in a forked repository
+* Forking an entire project (and subsequently managing upstream changes, etc.) is an unnecessarily tall order for something that can certainly be fixed with a simple, at-your-own-risk dependency override
+* This feature becomes more and more important with the number of gems growing. .NET has it. Java has it. Nothing bad happened to them because of it
+* Maintaining overridden version constraints within the Gemfile saves a significant amount of development cost when compared to maintaining a forked repository
+* Explicit syntax within the Gemfile with custom notes and obvious output messages when running `bundle install` can help to mitigate developers not knowing that a version conflict exists
+
+# Rationale and Alternatives
```

Comment:
Here is a good example of the problem in action: https://github.com/salesforce-marketingcloud/FuelSDK-Ruby/pull/115.

Reply:
I'm sorry, but I don't get it. But what makes it actual good example?

Reply:
Why would you prefer to maintain an entire fork of a gem just to update it's Gemfile, when you could just add an override in your own Gemfile?

The good things about overriding in your own Gemfile is that you are:
* `gem install` can explicitly show that a problematic gems dependencies are being forced to be a version that isn't technically supported by that gem. This wouldn't happen if you maintain a fork. 
* `gem install` can explicitly show that none of the problematic gem's dependencies' versions are being forced. If after a `gem update` this message is being show, it indicates that the original gem author has finally updated their problematic gem dependency version constraints, which means you can finally remove the override from your Gemfile. This wouldn't happen if you maintain a fork. You would need to constantly actually look at the original gem for updates. That is such a manual chore.

Reply:
> Why would you prefer to maintain an entire fork of a gem just to update it's Gemfile, when you could just add an override in your own Gemfile?

I would prefer to maintain an entire fork of a unmaintained gem or actually try to find out alternative to this gem to remove abandoned gem in all codebases I do maintain.

Reply:
> I would prefer to maintain an entire fork of a unmaintained gem or actually try to find out alternative to this gem to remove abandoned gem in all codebases I do maintain.

Ok, but why?

Reply:
> I would prefer to maintain an entire fork of a unmaintained gem or actually try to find out alternative to this gem to remove abandoned gem in all codebases I do maintain.

For the cases where that apply to us, we have found maintaining and tracking occasionally updated (but not fully abandoned) gems, ends up being a thankless chore that creates more problems than it solves. This of course doesn't mean it's right for everyone else.

This is also a common feature in other dependency managers:

- NPM (javascript): https://nodejs.org/en/blog/npm/managing-node-js-dependencies-with-shrinkwrap/
- NPM (javascript): https://www.npmjs.com/package/npm-force-resolutions
- Yarn (javascript): https://classic.yarnpkg.com/en/docs/selective-version-resolutions/
- Pip (python): https://pip.pypa.io/en/stable/user_guide/#constraints-files
- Maven (java): https://www.baeldung.com/maven-version-collision
- Gradle (java): https://docs.gradle.org/current/userguide/resolution_rules.html

The above cover the [top languages from Github](https://madnight.github.io/githut/#/pull_requests/2020/3) and all have a solution. I would argue this is not an esoteric need.

Reply:
I wrote up my view on the problem and options for solutions that are currently available, 1 that is this RFC and 1 other that is not available.

I hope this helps to frame the debate.

# Options for handling out of date gems

Out of date gems have been a problem for a long time. As the ruby ecosystem evolves and ages, they are and will become more common. This is especially prevalent in areas where the state of the art has moved on (for example SOAP).

Unfortunately, these 'stuck' gems often have dependencies on other more common and evolving gems. This would not be a problem if the un-evolved gems had loose dependency constraints. However, the recommendation for gem writers is to be relatively tight on constraints to ensure compatibility.

We now have stuck gems depending on other gems. Since only one version of a gem can be installed in bundler. This then constrains the use of newer gems that share a downstream gem dependency (but not the dependency version).

There are a number of options for handling these out of date gems. Here I list them out and highlight their pros and cons.

Finally, I give my recommendation for consideration.

Here are the many options that developers have to deal with out of date gems:

1. Create an issue
1. Create issue and develop a PR
1. Monkeypatch
1. Fork and use git path
1. Fork and use private registry
1. Fork and change name and alias name
1. Take over maintenance of the gem
1. Force version
1. Write your own

Here is a detailed description of each of them and their pros and cons:

## 1. Create an issue

Create an issue on the relevant repo and wait for the maintainer to handle it and release a new version. This and the next option are the happy path, the way things should work.

This option and the following option are usually the first port of call, since they are also the easiest.

### Pros

- Keeps gem up to date for the community
- Easy for the user

### Cons

- No help if the gem isn't being maintained
- Have to wait for maintainer to agree to and then make the change a priority
- Assumes gem is maintained in timely manner
- Maintainer has to do the work

## 2. Create issue and submit a PR

Create the issue but also provide a solution in the form of a PR. Again, this is the happy path, but isn't always available.

### Pros

- Keeps gem up to date for the community
- A ready made solution is in place so the maintainer has less work
- Relatively easy for some users

### Cons

- No help if the gem isn't being maintained
- Have to wait for maintainer to make it a priority
- Assumes gem is maintained in timely manner
- You have to do the work, and if you don't understand the code, figure it out.

## 3. Monkeypatch

Create a gem or an internal piece of code that overrides the buggy or out of date behavior in the gem that needs updating. An example of this is [Savon Fixes](https://rubygems.org/gems/savon_fixes) which I created.

### Pros

- Make the change without having to get the maintainers to do anything
- If a gem, can be provided to the community

### Cons

- Could cause missing out on gem updates
- Community doesn't move forward together because it could cause people to avoid proposing issues and changes to gems (options 1 and 2)
- Fragile (especially if a gem; a monkeypatch gem is *not* straightforward)
- Doesn't work for out of date dependencies as `gemspecs` can't be monkeypatched
- Depending on load order, can be very tricky to verify that it is working if deeply embedded in a chain of gems

## 4. Fork and use git path

This is the most common way of forking. Create the fork, make the patch and point your gem file at your git repo and the relevant commit/branch.

### Pros

- Can make targeted changes of whatever nature is needed

### Cons

- Maintaining a full code base (you have to do the work, and if you don't understand the code, figure it out)
- Requires discipline. I've seen additional functionality added to the forked gem, thus creating barriers to tracking upstream.
- Could cause missing out on gem updates
- Community doesn't move forward together because it could cause people to avoid proposing issues and changes to gems (options 1 and 2)
- Can't have all gems in a registry

## 5. Fork and use private registry

Create the fork, make the patch, add in your proprietary CI/CD and registry config/code and point your `Gemfile` at the internal registry.

### Pros

- Can make targeted changes of whatever nature is needed

### Cons

- Maintaining a full code base (you have to do the work, and if you don't understand the code, figure it out)
- Requires discipline. I've seen additional functionality added to the forked gem, thus creating barriers to tracking upstream.
- Could cause missing out on gem updates
- Community doesn't move forward together because it could cause people to avoid proposing issues and changes to gems (options 1 and 2)
- Having to add internal CI config/code to the public repo
- Having to create new version number to avoid clash with public version (so multiple sources aren't ambiguous)
- If there are intermediate gems, having to make sure they will accept the new version number

## 6. Public fork and change name

In this case you are forking the gem, and then changing it's name so that you can publish to `Rubygems` under the alternate name. Then your `Gemfile` points to the newly named gem.

### Pros

- Can use standard publishing mechanisms (`Rubygems`)
- Community can benefit from changes (if aware)

### Cons

- In order for intermediate gems in a dependency chain to use, would require `alias_gem_name` fuctionality to `Bundler` which is not currently available
- Multiple versions of same gem with similar names will exist
- Could cause missing out on original gem updates

## 7. Take over maintenance of the gem

In this case, become the maintainer of the gem as the previous maintainer no longer has the resources to spend the time on maintenance. Then you can make the necessary updates.

### Pros

- Keeps gem up to date for the community
- Gem is maintained again
- Strong way to give back to the community

### Cons

- Big task to take on
- Maintaining a full code base (you have to do the work, and if you don't understand the code, figure it out)
- When maintainer has disappeared, taking over maintainership is not easy
- Maintainer may not be willing to give up ownership, even as they no longer make changes. (I have an example of this.)
- To be a universal approach, requires more robust or frequent 'change of ownership' process for `Rubygems.org`

## 8. Force version

Add a capability `Bundler` which allows an option in a `Gemfile`. This option lets the user force the version of a dependency which then overrides that dependency in the `gemspecs` of all dependent gems.

### Pros

- Allows gem dependencies to be forced when the gem is still working (other than the out of date dependencies)
- Works on dependencies at any level in a dependency chain
- Javascript (nvm and yarn), Python (pip) and Java (Maven, Gradle), the top 3 Github languages, offer this capability

### Cons

- Could cause missing out on gem updates
- Community doesn't move forward together because it could cause people to avoid proposing issues and changes to gems (options 1 and 2)
- Change to `Bundler` api (though opt in)
- Does not solve cases where functionality/API changes are required

## 9. Write your own

This just means avoiding the 'stuck' gem entirely.

### Pros

- Full control

### Cons

- Not feasible in situations where the capabilities are not a strength of yours (in our case, SOAP)
- Does not solve problems for intermediate gems in a dependency chain
- If private, no benefit to or from the community
- If public, have to popularize your version so that all gem maintainers switch

## Recommendation

Given the complexity and how painful many of the options are, I believe adding 8) Force Version to `Bundler` is the best solution for the dependency side of these problems.

Additionally, I think the `Bundler` team should consider an `alias gem name` function that would allow users to redirect to a forked version of a gem which they can make public without having to fork entire chains of gems.

## References

Dependency managers offering override functionality:

- NPM (javascript): https://nodejs.org/en/blog/npm/managing-node-js-dependencies-with-shrinkwrap/
- NPM (javascript): https://www.npmjs.com/package/npm-force-resolutions
- Yarn (javascript): https://classic.yarnpkg.com/en/docs/selective-version-resolutions/
- Pip (python): https://pip.pypa.io/en/stable/user_guide/#constraints-files
- Maven (java): https://www.baeldung.com/maven-version-collision
- Gradle (java): https://docs.gradle.org/current/userguide/resolution_rules.html

- Top languages in Github: https://madnight.github.io/githut/#/pull_requests/2020/3
- Google search for 'bundler override gem dependency  site:stackoverflow.com' returns 1,360 results: https://www.google.com/search?q=bundler+override+gem+dependency+site:stackoverflow.com